src/main/java/org/candlepin/subscriptions/capacity/CapacityReconciliationController.java

```diff
@@ -0,0 +1,115 @@
+import org.candlepin.subscriptions.capacity.files.ProductWhitelist;
+import org.candlepin.subscriptions.db.OfferingRepository;
+import org.candlepin.subscriptions.db.SubscriptionCapacityRepository;
+import org.candlepin.subscriptions.db.model.*;
+import org.springframework.stereotype.Component;
+
+@Component
+@Slf4j
+public class CapacityReconciliationController {
+
+  private final OfferingRepository offeringRepository;
+  private final ProductWhitelist productWhitelist;
+  private final CapacityProductExtractor productExtractor;
+  private final SubscriptionCapacityRepository subscriptionCapacityRepository;
+  private final Counter capacityRecordsCreated;
+  private final Counter capacityRecordsUpdated;
+  private final Counter capacityRecordsDeleted;
+
+  public CapacityReconciliationController(
+      OfferingRepository offeringRepository,
+      ProductWhitelist productWhitelist,
+      CapacityProductExtractor productExtractor,
+      SubscriptionCapacityRepository subscriptionCapacityRepository,
+      MeterRegistry meterRegistry) {
+    this.offeringRepository = offeringRepository;
+    this.productWhitelist = productWhitelist;
+    this.productExtractor = productExtractor;
+    this.subscriptionCapacityRepository = subscriptionCapacityRepository;
+    capacityRecordsCreated = meterRegistry.counter("rhsm-subscriptions.capacity.records_created");
+    capacityRecordsUpdated = meterRegistry.counter("rhsm-subscriptions.capacity.records_updated");
+    capacityRecordsDeleted = meterRegistry.counter("rhsm-subscriptions.capacity.records_deleted");
+  }
+
+  public void reconcileCapacityForSubscription(Subscription subscription) {
+
+    Collection<SubscriptionCapacity> newCapacities = mapSubscriptionToCapacities(subscription);
+    reconcileSubscriptionCapacities(
+        newCapacities, subscription.getSubscriptionId(), subscription.getSku());
+  }
+
+  private Collection<SubscriptionCapacity> mapSubscriptionToCapacities(Subscription subscription) {
+
+    Offering offering = offeringRepository.getById(subscription.getSku());
+    Set<String> products =
+        productExtractor.getProducts(
+            offering.getProductIds().stream().map(Object::toString).collect(Collectors.toSet()));
+
+    return products.stream()
+        .map(product -> from(subscription, offering, product))
+        .collect(Collectors.toList());
+  }
+
+  private void reconcileSubscriptionCapacities(
+      Collection<SubscriptionCapacity> newCapacities, String subscriptionId, String sku) {
+
+    Collection<SubscriptionCapacity> toSave = new ArrayList<>();
+    Map<SubscriptionCapacityKey, SubscriptionCapacity> existingCapacityMap =
+        subscriptionCapacityRepository.findByKeySubscriptionId(subscriptionId).stream()
+            .collect(Collectors.toMap(SubscriptionCapacity::getKey, Function.identity()));
+
+    if (productWhitelist.productIdMatches(sku)) {
+      newCapacities.forEach(
+          newCapacity -> {
+            toSave.add(newCapacity);
```

Comment:
It looks like this method would either save all or delete all. Do we need to consider `SubscriptionCapacity`s that exist and are unchanged from their old value, and leave them untouched?

Reply:
I translated whatever is being done today to reconcile capacities in the pool ingress controller. The idea is to add or update any new capacities and remove stale ones. I could add a check to leave unchanged capacities untouched, but imo it adds unnecessary complexity to the implementation with not a lot of value. The current approach assumes that the latest state is the correct state. Does that make sense?

Reply:
:+1: from my PoV